rollover.py

```diff
@@ -0,0 +1,300 @@
+        return
+
+    #
+    # learn about and sort the instances by availability zone
+    #
+    ecs_descriptions = ecs_client.describe_instances(selected_instances)
+    ec2_to_ecs_id = dict([(v['ec2InstanceId'], k) for k, v in ecs_descriptions.items()])
+    ec2_ids = ec2_to_ecs_id.keys()
+    ec2_descriptions = ec2_client.describe_instances(ec2_ids)
+
+    # sort and warn by availability zones
+    asg_instances = asg.describe_instances()
+    sorted_ec2_ids = sort_and_check_availability_zones(ec2_descriptions,
+                                                       asg_instances)
+    if not sorted_ec2_ids:
+        return
+
+    #
+    # Query all services and tasks on each ECS instance
+    #
+    sys.stdout.write("Querying ECS services...")
+    sys.stdout.flush()
+    service_ids = ecs_client.list_services()
+    service_descriptions = ecs_client.describe_services(service_ids)
+    service_events = map_service_events(service_descriptions)
+
+    task_ids = ecs_client.list_tasks()
+    task_descriptions = ecs_client.describe_tasks(task_ids)
+    ecs_instance_services = map_instance_services(service_descriptions,
+                                                  task_descriptions)
+    asg_ids = [asg_instance['InstanceId'] for asg_instance in asg_instances]
+    sorted_asg_ids = [ec2_id for ec2_id in sorted_ec2_ids if ec2_id in asg_ids]
+    print "done"
+
+    #
+    # Iterate through each instance
+    #
+    for ec2_id in sorted_asg_ids:
+        ecs_id = ec2_to_ecs_id[ec2_id]
+        print "Preparing to remove %s (%s)" % (ecs_id, ec2_id)
+
+        #
+        # Remove ECS instance from scaling group
+        #
+        sys.stdout.write("Removing EC2 instance %s from scaling group and waiting for replacements..." % (ec2_id))
+        sys.stdout.flush()
+        if not args.noop:
+            if args.scale_down:
+                asg.detach_instances([ec2_id], scale_down=True)
+            else:
+                asg.detach_instances_and_wait([ec2_id])
+        print "done"
+
+        #
+        # De-register instances from ECS
+        #
+        sys.stdout.write("De-registering instance %s from ECS..." % (ecs_id))
+        sys.stdout.flush()
+        if not args.noop:
+            ecs_client.deregister_container_instance(ecs_id)
+        print "done"
+
+        #
+        # Wait for task migrations
+        #
+        sys.stdout.write("Rolling over services from %s ..." % (ecs_id))
+        sys.stdout.flush()
+        if not args.noop:
+            for service_id in ecs_instance_services[ecs_id]:
+                last_event = service_events[service_id][-1]
+                event = ecs_client.wait_for_service_steady_state(service_id,
+                                                                 last_event)
+                # push the new event into the list for that service so that
+                # the next instance doesn't confuse this event for its own
+                service_events[service_id].append(event)
```

Comment:
yes, ecs service events are tracked per service, which means each time we remove an instance, it generates more events, This code is simply updating the events instead of calling describe_services again.